### PR TITLE
Rule Fixes

### PR DESCRIPTION
**Pull request by TrevorBrunette**
Simple rule fixes:
* Council needs 3 or more players to grant life
* Bounty can only happen if the council has enough coffer
* Code blocks replaced with lambdas where applicable

### CODE REVIEW

File: src/main/java/pro/trevor/tankgame/rule/impl/version3/Ruleset.java [hunk 1]
```diff
@@ -72,21 +72,7 @@ public void registerPlayerRules(RulesetDescription ruleset) {
...
-                    int bounty = toType(n[1], Integer.class);
-                    assert Attribute.COFFER.unsafeFrom(c) >= bounty;
-                    t.setBounty(t.getBounty() + bounty);
-                    Attribute.COFFER.to(c, Attribute.COFFER.unsafeFrom(c) - bounty);
-                    Attribute.CAN_BOUNTY.to(c, false);
-                },
-                UnitRange.ALL_LIVING_TANKS,
-                new DiscreteIntegerRange("bounty", 1, 5)));
+        playerRules.put(Council.class, PlayerRules.GetRuleCofferCostGrantLife(15, 3));
+        playerRules.put(Council.class, PlayerRules.GetRuleCofferCostBounty(1, 5));
```
Comment (ryan3r) [inline src/main/java/pro/trevor/tankgame/rule/impl/version3/Ruleset.java:76]: Down with the code duplication!  We love to see it.

File: src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java [hunk 2]
```diff
@@ -166,6 +165,27 @@ public static PlayerActionRule<Council> GetRuleCofferCostGrantLife(int cost) {
                 UnitRange.ALL_TANKS);
     }
 
+    public static PlayerActionRule<Council> GetRuleCofferCostBounty(int lowerBound, int upperBound) {
```
Comment (ryan3r) [inline src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java:168]: Please add a unit test for bounty.  You can probably copy the grant life tests.

**Review by ryan3r — CHANGES_REQUESTED** [review]

File: src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java [hunk 3]
```diff
@@ -166,6 +165,27 @@ public static PlayerActionRule<Council> GetRuleCofferCostGrantLife(int cost) {
...
+        assert lowerBound >= 0 && upperBound >= lowerBound;
+        return new PlayerActionRule<>(PlayerRules.ActionKeys.BOUNTY,
+                (s, c, n) -> {
```
[the remark this answers is not in the document]
Comment (ryan3r) [reply src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java:171]: Note: I'm not requiring this just suggesting

File: src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java [hunk 4]
```diff
@@ -145,12 +143,13 @@ public static PlayerActionRule<Council> GetCofferCostStimulusRule(int cost) {
...
     }
 
-    public static PlayerActionRule<Council> GetRuleCofferCostGrantLife(int cost) {
+    public static PlayerActionRule<Council> GetRuleCofferCostGrantLife(int cost, int minimumCouncillors) {
         return new PlayerActionRule<>(
                 PlayerRules.ActionKeys.GRANT_LIFE,
                 (s, c, n) -> {
                     GenericTank t = toType(n[0], GenericTank.class);
-                    return Attribute.DEAD.in(t) && Attribute.DURABILITY.in(t) && Attribute.COFFER.unsafeFrom(c) >= cost;
+                    return Attribute.DEAD.in(t) && Attribute.DURABILITY.in(t) && Attribute.COFFER.unsafeFrom(c) >= cost
```
[the remark this answers is not in the document]
Comment (TrevorBrunette) [reply src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java:151]: Tanks do not have to be dead for them to be able to be granted life.

[the remark this answers is not in the document]
Comment (ryan3r) [reply src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java:151]: Oh, then why is there a check for dead?

[the remark this answers is not in the document]
Comment (TrevorBrunette) [reply src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java:151]: Grant life is checking to ensure that the generic tank is capable of being dead (Attribute.DEAD.in is not Attribute.DEAD.from)

**Review by ryan3r — APPROVED** [review]

**Pull request merged**